File: StriverA2Z/Java/LinkedLists/Questions/SinglyLinkedList/CopyListWithRandomPointers.java
class CopyNode {
    int val;
    CopyNode next;
    CopyNode random;
    CopyNode(int value) {
        val = value;
        next = null;
        random = null;
    }
}
public class CopyListWithRandomPointers {
    static CopyNode copyRandomList(CopyNode head) {
        CopyNode temp = head;
        //step 1
        while(temp != null) {
            CopyNode newNode = new CopyNode(temp.val);
            newNode.next = temp.next;
            temp.next = newNode;
            temp = temp.next.next;
        }
        //step 2
        CopyNode itr = head;
        while(itr != null) {
            if(itr.random != null)
                itr.next.random = itr.random.next;
            itr = itr.next.next;
        }
        //step 3
        CopyNode dummy = new CopyNode(0);
        itr = head;
        temp = dummy;
        CopyNode fast;
        while(itr != null) {
            fast = itr.next.next;
            temp.next = itr.next;
            itr.next = fast;
            temp = temp.next;
            itr = fast;
        }
        return dummy.next;
    }

    static void printList(CopyNode head) {
        while(head != null) {
            System.out.print(head.val+":");
            if(head.next != null)
                System.out.print(head.next.val);
            else
                System.out.print("NULL");
            if(head.random != null)
                System.out.print(","+head.random.val);
            else
                System.out.print(",NULL");
            System.out.println();
            head = head.next;
        }
    }

    public static void main(String args[]) {
        CopyNode head = null;

        CopyNode node1 = new CopyNode(1);
        CopyNode node2 = new CopyNode(2);
        CopyNode node3 = new CopyNode(3);
        CopyNode node4 = new CopyNode(4);

        head = node1;
        head.next = node2;
        head.next.next = node3;
        head.next.next.next = node4;

        head.random = node4;
        head.next.random = node1;
        head.next.next.random = null;
        head.next.next.next.random = node2;

        System.out.println("Original list:(current node:node pointed by next pointer, node pointed by random pointer)");
        printList(head);

        System.out.println("Copy list:(current node:node pointed by next pointer,node pointed by random pointer)");
        CopyNode newHead = copyRandomList(head);
        printList(newHead);
    }

}

File: StriverA2Z/C++/6_LinkedList/Questions/SinglyLinkedList/copy_list_with_random_pointers.cpp
#include<bits/stdc++.h>

using namespace std;

class Node {
public:
    int val;
    Node *next;
    Node *random;

    Node(int value) {
        val = value;
        next = nullptr;
        random = nullptr;
    }
};

Node *copyRandomList(Node *head) {
    Node *temp = head;
    while (temp) {
        Node *copy = new Node(temp->val);
        copy->next = temp->next;
        temp->next = copy;
        temp = temp->next->next;
    }
    temp = head;
    while (temp) {
        Node *copy = temp->next;
        if (temp->random) copy->random = temp->random->next;
        else copy->random = nullptr;
        temp = temp->next->next;
    }
    Node *dNode = new Node(-1);
    Node *res = dNode;
    temp = head;
    while (temp) {
        res->next = temp->next;
        temp->next = temp->next->next;
        res = res->next;
        temp = temp->next;
    }
    return dNode->next;
}

void printList(Node *head) {
    while (head != nullptr) {
        cout << head->val << ':';
        head->next != nullptr ? cout << head->next->val : cout << "NULL";
        head->random != nullptr ? cout << "," << head->random->val : cout << ",NULL";
        cout << endl;
        head = head->next;
    }
}

int main() {
    Node *head = nullptr;

    Node *node1 = new Node(1);
    Node *node2 = new Node(2);
    Node *node3 = new Node(3);
    Node *node4 = new Node(4);

    head = node1;
    head->next = node2;
    head->next->next = node3;
    head->next->next->next = node4;

    head->random = node4;
    head->next->random = node1;
    head->next->next->random = nullptr;
    head->next->next->next->random = node2;

    cout << "Original list:(current node:node pointed by next pointer,node pointed by random pointer)\n";
    printList(head);

    cout << "Copy list:(current node:node pointed by next pointer,node pointed by random pointer)\n";
    Node *newHead = copyRandomList(head);
    printList(newHead);
    return 0;
}